[#1496] Image conversion

cpiber opened the pull request:
- Optional conversion on download
  - This conversion is only done on actual download (via queue), not on cache
  - Configurable per image mime type, with a possible default conversion
- Query parameter to convert in-memory on request
  - If page already downloaded (or local), converts that in-memory
  - Otherwise downloads the page without conversion, then converts in-memory

The main reason for this design decision is that `Page` is responsible both for interactive requests and downloads, so supporting both download conversion and user-provided conversion is more complex. It also makes reading without downloads slower, as conversion takes noticeable time.

The current configuration for download conversions is done via an object. The config library converts this to `server.downloadConversions.default = "..."` and `server.downloadConversions["image/webp"] = "..."`, hence the special case in `ConfigManager`. Maybe we can find a better format, which is better supported?

Closes 1496

cpiber committed:
[#1496] First conversion attempt

cpiber committed:
[#1496] Configurable conversion

cpiber committed:
Fix: allow nested configs (map)

cpiber committed:
[#1496] Support explicit `none` conversion

cpiber committed:
Use MimeUtils for provided download

cpiber committed:
[1496] Support image conversion on load for downloaded images

cpiber committed:
Lint

cpiber committed:
[#1496] Support conversion on fresh download as well

Previous commit was only for already downloaded images, now also for
fresh and cached

cpiber committed:
[#1496] Refactor: Move where conversion for download happens

Syer10 commented:
I agree that the config format is getting in out way a bit. I would be open to alternatives.

I have had good experiences with TOML in the past, and there is a support TOML library for Kotlin here https://github.com/orchestr7/ktoml


Do you want to take a look at this? I can look at this in my own time if not. Let me know what your plan is, we can move forward with this PR as it is for now.

cpiber commented:
> I agree that the config format is getting in out way a bit. I would be open to alternatives.
> 
> I have had good experiences with TOML in the past, and there is a support TOML library for Kotlin here [orchestr7/ktoml](https://github.com/orchestr7/ktoml?rgh-link-date=2025-07-07T21%3A22%3A11Z), though I'm not sure how comments would work with that or other formats.

I was more referring to the format of the conversion mapping. I hadn't thought of completely rewriting the config system.

> Do you want to take a look at this? I can look at this in my own time if not. Let me know what your plan is, we can move forward with this PR as it is for now.

For now, I don't think I want to work on migrating/rewriting the config system. That might be interesting for the future, but at the moment I don't have that much time and my vacation is coming up, where I want to reduce contributions

cpiber commented:
We could also simplify what can be configured.

I currently have it set up as a map (object) from input mime type to output mime type, but I think realistically users will want to convert their entire library to some format, or not at all. Maybe we could have a whitelist of formats to be converted (e.g. convert "image/jpeg" and "image/png" to "image/webp"), that should cover most usecases

Syer10 committed:
Rewrite config handling, improve custom types

Syer10 committed:
Lint

Syer10 committed:
Add format to pages mutation

Syer10 committed:
Lint

cpiber commented:
I had actually planned options originally, don't know why I forgot about it, will do!

Syer10 committed:
Standardize url encode

Syer10 committed:
Lint

cpiber committed:
Config: Allow additional conversion parameters

cpiber committed:
Implement conversion quality parameter

cpiber committed:
Lint

cpiber commented:
[ImageWriteParam](https://docs.oracle.com/javase/8/docs/api/javax/imageio/ImageWriteParam.html) only supports quality and tiling. Tiling does not seem useful (and I believe is only sensible for JPEG), so quality is the only setting for now.

I've also added an example to the reference config, and updated config manager to copy that over. With `config.entrySet()`, only non-null entries are listed (recursively). But an empty object (the default for conversions) is essentially null, so it was never detected. Instead manually iterate the first two levels (the expected depth) and use that as basis, since that also includes empty objects.

Note that this format would also work, since any JSON is allowed:

```
server.downloadConversions = {
  "image/webp": {
    target = "image/jpeg"
  }
}
```

cpiber committed:
Implement a conversion util to allow fallback readers

Syer10 committed:
Add downloadConversions to api and backup, fix updateValue issues

Syer10 committed:
Lint

Syer10 committed:
Minor cleanup

Syer10 committed:
Merge branch 'master' into image-convert

Syer10 committed:
Update libs.versions.toml
